src/frontend/metrics/AppDiagnosticMetrics.ts

```diff
@@ -0,0 +1,204 @@
+
+    subscriptionCleanupFns.push(
+      NetInfo.addEventListener(state => {
+        this.#isOnline = Boolean(state.isInternetReachable);
+        this.#update();
+      }),
+    );
+
+    const appStateSubscription = AppState.addEventListener(
+      'change',
+      this.#update,
+    );
+    subscriptionCleanupFns.push(() => appStateSubscription.remove());
+
+    // We could be smarter about this. For example, we could set the timer for
+    // beginning of the next day. However, timeouts are a bit unreliable (for
+    // example, if the app goes into the background), so we just check
+    // periodically. This also simplifies the code a bit.
+    setInterval(this.#update, CHECK_INTERVAL);
+
+    this.#subscriptionCleanupFns = subscriptionCleanupFns;
+  }
+
+  #removeListeners(): void {
+    const subscriptionCleanupFns = this.#subscriptionCleanupFns ?? [];
+    for (const fn of subscriptionCleanupFns) fn();
+  }
+
+  /**
+   * Update the queue and send metrics as needed.
+   *
+   * Should only be run once at a time. See `this.#update` for a queued alternative.
+   */
+  async #doUpdate(): Promise<void> {
+    if (!this.#isEnabled) {
+      storage.delete(STORAGE_KEY);
+      return;
+    }
+
+    let queue = loadQueueFromStorage();
+
+    const now = new Date();
+
+    const numberOfReportsBeforeTruncation = queue.reports.length;
+    queue = truncateReportsByTime(queue, now);
+    const numberOfReportsAfterTruncation = queue.reports.length;
+    let hasChangedQueue =
+      numberOfReportsBeforeTruncation !== numberOfReportsAfterTruncation;
+
+    if (!hasReportForToday(queue)) {
+      queue = {
+        ...queue,
+        reports: [...queue.reports, await generateAppDiagnosticMetricsData()],
+      };
+      hasChangedQueue = true;
+    }
+
+    if (hasChangedQueue) saveQueueToStorage(queue);
+
+    const newestReport = last(queue.reports);
+
+    const shouldSendMetrics =
+      this.#isEnabled &&
+      this.#isOnline &&
+      AppState.currentState === 'active' &&
+      !!newestReport;
+    if (!shouldSendMetrics) return;
+
+    try {
+      const newestReportDate = parseIsoUtc(newestReport.dateGenerated);
+      assert(newestReportDate, 'Expected report to be generated');
+
+      await sendMetricsData({
+        ...getMetricsRequestInfo(),
+        dataToSend: {
+          type: 'app diagnostics v1',
+          monthlyDeviceHash: await getMonthlyHash(
+            'app diagnostics',
+            getMetricsDeviceId(),
+            // We prefer the date from the newest report, not `new Date()`.
+            // These will usually be the same, but it's possible for a user to
+            // send metrics right at the end of the month which would generate
+            // a new monthly hash, letting us track users across months.
+            newestReportDate,
+          ),
+          reports: queue.reports,
+        },
+      });
+
+      queue = updateQueueHighWatermark(queue);
+      saveQueueToStorage(queue);
```

Comment:
Why keep reports in the queue after a successful send?

Reply:
We don't. `updateQueueHighWatermark` returns an object like `{highWatermark: '2024-08-12', reports: []}`. That function probably needs to be renamed for clarity.